### ISSUE TITLE
[V3] Hue Color Temperature in Group causes invalid status

### ISSUE THREAD

**Issue by nougad** [openhab/openhab-alexa#210]
**tl;dr: I think the skill should not forward `NULL` state of `ColorTemperatureController.colorTemperatureInKelvin` when used in a group since it causes bugs with Hue**

I defined a Group with Color and ColorTemperature for a hue bulb as described here: https://github.com/openhab/openhab-alexa/issues/174#issuecomment-475878390

```
Group Kitchen1 "Küche 1" <colorwheel> ( Kitchen, gLight ) {alexa="Endpoint.LIGHT"}
Color
  KChe1_Color
  ( Kitchen1 )
  {channel="hue:0210:00...3:kueche1:color", alexa="BrightnessController.brightness,PowerController.powerState,ColorController.color"}
Dimmer
  KChe1_ColorTemperature
  ( Kitchen1 )
  {channel="hue:0210:00...3:kueche1:color_temperature", alexa="ColorTemperatureController.colorTemperatureInKelvin"}
```

But whenever I update the color via Alexa, OpenHab or hue, Alexa will complain the `Device is unresponsive`.

The reason is the Hue logic to define color:
* *White Bulbs* have only one `Dimmer` in OpenHab to set the brightness
* *White Ambiance Bulbs* have two `Dimmer` in OpenHab. One for brightness and one for ColorTemperature. Since both values are unrelated to each other it works fine. You can change the brightness and Temperature independently.
* *White and Color Ambiance Bulbs* have one `Color` for brightness and color and one `Dimmer` for Temperature. The Hue app shows then two modes. One for normal Temperature selection and one for color selection.
  * If the temperature is selected the value is translated to a color setting and both, the `Dimmer` and the `Color` item have updated values.
  * Same if the brightness is modified. That will keep the temperature and only translate the color to a new value.
  * The problem is if the color is changed to a value which can't be represented by the white temperature. In that case **Hue will only modify the color and set the `Dimmer` state to `NULL`**.

Having a `NULL` value for Temperature in the group will then cause Alexa to show the device as unresponsive.

I think this is not perfectly solved in OpenHab since the brightness is bundled with the Color and there is no way to change the brightness without loosing the temperature setting since hue will switch to color-mode and set temperature to `NULL`. But in combination with the Alexa skill this is a Bug.

I didn't find a way to turn on request/response body logging. [openhabcloud CloudClient](https://github.com/wallaroo/openhab2-addons/blob/51e618ddcacc182d0758247c33e9b7fb88411f8f/addons/io/org.openhab.io.openhabcloud/src/main/java/org/openhab/io/openhabcloud/internal/CloudClient.java) only logs headers and Jetty HttpClient does not log any useful output either. So it's difficult to debug what goes over the wire.

I did not find a way to overwrite the state of Temperature to avoid setting it to `NULL`. I can do a rule on Item changed and set it to another value but then I override the Color value. The only workaround is to not use Temperature on Hue Color Bulbs.

It would be great if there is a group with two items including Color and Temperature and the temperature is `NULL` if the skill could ignore the temperature. Not sure if that's possible.

# Examples:

## changing the temperature:

```
14:57:36.980 [INFO ] [smarthome.event.ItemStateEvent       ] - KChe1_Color updated to 279,15,100 
14:57:36.983 [INFO ] [smarthome.event.ItemStateEvent       ] - KChe1_ColorTemperature updated to 56  
14:57:36.985 [INFO ] [smarthome.event.ItemStateChangedEvent] - KChe1_Color changed from 7,25,100 to 279,15,100 
14:57:36.986 [INFO ] [smarthome.event.ItemStateChangedEvent] - KChe1_ColorTemperature changed from NULL to 56  
```

## changing the color:

```
16:05:45.873 [INFO ] [smarthome.event.ItemStateEvent       ] - KChe1_Color updated to 13,55,37
16:05:45.875 [INFO ] [smarthome.event.ItemStateEvent       ] - KChe1_ColorTemperature updated to NULL
16:05:45.877 [INFO ] [smarthome.event.ItemStateChangedEvent] - KChe1_Color changed from 81,39,37 to 13,55,37
```

## turning the light off:

```
16:06:36.007 [INFO ] [smarthome.event.ItemStateEvent       ] - KChe1_Color updated to 13,55,0
16:06:36.008 [INFO ] [smarthome.event.ItemStateEvent       ] - KChe1_ColorTemperature updated to NULL
16:06:36.009 [INFO ] [smarthome.event.ItemStateChangedEvent] - KChe1_Color changed from 13,55,37 to 13,55,0
```

## API representation

```
$ curl -X GET --header "Accept: application/json" "http://.../rest/items/Kitchen1"
{
    "members": [
        {
            "link": "http://.../rest/items/KChe1_Color",
            "state": "13,55,0",
            "editable": false,
            "type": "Color",
            "name": "KChe1_Color",
            "tags": [],
            "groupNames": [
                "Kitchen1"
            ]
        },
        {
            "link": "http://.../rest/items/KChe1_ColorTemperature",
            "state": "NULL",
            "editable": false,
            "type": "Dimmer",
            "name": "KChe1_ColorTemperature",
            "tags": [],
            "groupNames": [
                "Kitchen1"
            ]
        }
    ],
    "link": "http://.../rest/items/Kitchen1",
    "state": "NULL",
    "editable": false,
    "type": "Group",
    "name": "Kitchen1",
    "label": "K\u00fcche 1",
    "category": "colorwheel",
    "tags": [],
    "groupNames": [
        "Kitchen",
        "gLight"
    ]
}
```

**Comment by jsetton** [comment]
@nougad thanks for the detailed information.

> But whenever I update the color via Alexa, OpenHab or hue, Alexa will complain the `Device is unresponsive.`

I just run a quick test. I assume the issue you are seeing is from the Alexa app standpoint which is trying and failing to show the proper current state. Voice commands (set/adjust color and temperature) are working fine from my testing when the temperature has a `NULL` state.

> Having a NULL value for Temperature in the group will then cause Alexa to show the device as unresponsive.

So, the [Alexa skill](https://github.com/openhab/openhab-alexa/blob/master/lambda/smarthome/alexa/directive.js#L154) will determine any linked items with a state of `NULL` or `UNDEF` as undefined state and therefore ultimately generate a not responsive error since it is not able to report the state of that specific item part of the group.

> I think this is not perfectly solved in OpenHab since the brightness is bundled with the Color and there is no way to change the brightness without loosing the temperature setting since hue will switch to color-mode and set temperature to `NULL`

Is the Hue binding intentionally setting the temperature item to `NULL` when the color is updated? If so, we might have to make a change that would differentiate `NULL` and `UNDEF` OH states where the former would just not be reported back to Alexa but the overall state of the group endpoint would still be considered responsive while with the latter, we would keep the current implementation.

**Comment by nougad** [comment]
> I just run a quick test. I assume the issue you are seeing is from the Alexa app standpoint which is trying and failing to show the proper current state. Voice commands (set/adjust color and temperature) are working fine from my testing when the temperature has a NULL state.

yes, that's correct. I'm talking about app-state.

> Is the Hue binding intentionally setting the temperature item to NULL when the color is updated?

I suspect this is the causing line: https://github.com/openhab/openhab2-addons/blob/53342ee89be902892a826696a72e9a1620c551bd/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/HueLightHandler.java#L472 - No idea if that's on purpose. The hue internal API (unrelated to OH) defines a operating state:

> colormode:  Indicates the color mode in which the light is working, this is the last command type it received. Values are “hs” for Hue and Saturation, “xy” for XY and “ct” for Color Temperature. This parameter is only present when the light supports at least one of the values.

Since hue itself does not have the concept of Color AND Temperature so the mapping to two items in OH is arbitrary and the value when switching to Color is undefined.

**Comment by jsetton** [comment]
> I suspect this is the causing line: https://github.com/openhab/openhab2-addons/blob/53342ee89be902892a826696a72e9a1620c551bd/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/HueLightHandler.java#L472 - No idea if that's on purpose

Thanks for confirming. I did some digging and basically the Alexa API expects either the [color](https://developer.amazon.com/docs/device-apis/alexa-colorcontroller.html#properties) or [temperature](https://developer.amazon.com/docs/device-apis/alexa-colortemperaturecontroller.html#properties) properties to be reported depending on the color mode. Currently the skill is reporting both which is actually messing up the app state even if both values are defined.

Looking at the various bindings supporting color temperatures (e.g. Hue, LIFX, Milight & Yeelight) implementations, it appears that they all used different rules with some of them not differentiating color mode clearly.

In order to fix the current color temperature implementation, I am planning to add the following:
- ability to limit supported color temperature range via metadata parameter and binding preset-values
- ability to force color saturation to `0` when setting the color temperature to a specific shade of white via metadata boolean parameter
- add logic to determine color mode and return proper properties response back to Alexa

From the research I have done so far, here are the conditions I plan to use to determine the color mode is `on` when a group endpoint has both color and color temperature interfaces defined:
- color temperature set to `NULL` for Dimmer; `0` or `NULL` for Number
- color saturation not set to `0`

@nougad Unfortunately, I don't have a hue bulb to test with. Can you please confirm what happen when you just change the color temperature from the hue app perspective? Does it switch to a shade of white or mix it with the current color setting? In the event of the former, I am curious to see what the color state gets updated to and more specifically if the color saturation (h, **s**, b) is changed on the OH side.
Same question when setting your hue color bulbs to a shade of white via Alexa (_set hue light to Soft White_), basically sending a command to the color temperature channel on the OH side?

**Comment by nougad** [comment]
Thanks for looking into this! And thanks a lot for the change!!

Hue has two modes. One for color and one for temperature. Only one mode can be activated at the same time.

Here is the raw output from hue API:

# lamp off 
 
## hue api 
 
``` 
curl -s -H "Accept: application/json" http://${HUEHOST}/api/${HUEUSER}/lights/5 | jq .state· 
{ 
  "on": false, 
  "bri": 254, 
  "hue": 63574, 
  "sat": 253, 
  "effect": "none", 
  "xy": [ 
    0.62, 
    0.2923 
  ], 
  "ct": 153, 
  "alert": "select", 
  "colormode": "xy", 
  "mode": "homeautomation", 
  "reachable": true 
} 
``` 
 
# lamp on - color mode 
 
## openhab logs 
 
``` 
17:22:15.453 [INFO ] [smarthome.event.ItemStateEvent       ] - KChe1_Color updated to 345,100,100 
17:22:15.454 [INFO ] [smarthome.event.ItemStateEvent       ] - KChe1_ColorTemperature updated to NULL 
17:22:15.455 [INFO ] [smarthome.event.ItemStateChangedEvent] - KChe1_Color changed from 345,100,0 to 345,100,100 
``` 
 
## hue api 
 
``` 
{ 
  "on": true, 
  "bri": 254, 
  "hue": 63574, 
  "sat": 253, 
  "effect": "none", 
  "xy": [ 
    0.62, 
    0.2923 
  ], 
  "ct": 153, 
  "alert": "select", 
  "colormode": "xy", 
  "mode": "homeautomation", 
  "reachable": true 
} 
``` 
 
# lamp on - temperature mode 
 
## openhab logs 
 
``` 
17:22:45.518 [INFO ] [smarthome.event.ItemStateEvent       ] - KChe1_Color updated to 349,100,100 
17:22:45.519 [INFO ] [smarthome.event.ItemStateEvent       ] - KChe1_ColorTemperature updated to 89 
17:22:45.520 [INFO ] [smarthome.event.ItemStateChangedEvent] - KChe1_Color changed from 345,100,100 to 349,100,100 
17:22:45.521 [INFO ] [smarthome.event.ItemStateChangedEvent] - KChe1_ColorTemperature changed from NULL to 89 
``` 
 
## hue api 
 
``` 
{ 
  "on": true, 
  "bri": 254, 
  "hue": 13249, 
  "sat": 209, 
  "effect": "none", 
  "xy": [ 
    0.5095, 
    0.4152 
  ], 
  "ct": 462, 
  "alert": "select", 
  "colormode": "ct",              ##### mode switched 
  "mode": "homeautomation", 
  "reachable": true 
} 
``` 
 
# lamp off 
 
## openhab logs 
 
``` 
17:23:35.634 [INFO ] [smarthome.event.ItemStateEvent       ] - KChe1_Color updated to 72,83,0 
17:23:35.635 [INFO ] [smarthome.event.ItemStateEvent       ] - KChe1_ColorTemperature updated to 89 
17:23:35.636 [INFO ] [smarthome.event.ItemStateChangedEvent] - KChe1_Color changed from 72,83,100 to 72,83,0 
``` 
 
## hue api 
 
``` 
{ 
  "on": false, 
  "bri": 254, 
  "hue": 13249, 
  "sat": 209, 
  "effect": "none", 
  "xy": [ 
    0.5095, 
    0.4152 
  ], 
  "ct": 462, 
  "alert": "select", 
  "colormode": "ct", 
  "mode": "homeautomation", 
  "reachable": true 
} 
```

**Comment by jsetton** [comment]
Thanks for the detailed information. These values are in line to what I expected. The change I submitted will cover the issue you reported.

**Comment by nougad** [comment]
sorry to ping here again. I'm using v3 beta and I finally gave it a try but I still get `Device is unresponsive` with the following configuration:

```
Group:Switch:OR(ON, OFF) Bambus "Bambus" <colorwheel> (LivingRoom) {alexa="Endpoint.LIGHT"} 
Color WohnzimmerBambus_Color 
  ( Bambus ) 
  { channel="hue:0210:0...3:wozi-spot-bambus:color", alexa="Lighting" } 
Dimmer WohnzimmerBambus_ColorTemperature 
  ( Bambus ) 
  { channel="hue:0210:0...3:wozi-spot-bambus:color_temperature", alexa="ColorTemperature" [binding="hue"] } 
```
I'm doing something wrong or is it not deployed to V3 beta yet? I see it's only pushed to `master` branch. I'm wondering if I leave the beta, would it then work?

Without the `Dimmer` part it works fine (but without the dimmer functionality of course).

**Comment by jsetton** [comment]
> I'm doing something wrong or is it not deployed to V3 beta yet? I see it's only pushed to master branch. I'm wondering if I leave the beta, would it then work?

I don't have the visibility of the V3 beta skill but I am fairly certain that this change wasn't push to the beta version. So that would explain your observation. Unfortunately, you would have to wait for the official skill to be updated to V3 as we are in the process of getting it approved by Amazon, unless you are willing to deploy your own skill.

**Comment by nougad** [comment]
I was able to test the change after the [latest release](https://community.openhab.org/t/alexa-smart-home-skill-v3-is-now-live/79848).

I can switch between temperature and color mode and the App works just fine. EXCEPT when I select a color temperature the previous color is still reported. That makes the App to show the wrong color.

example:

* Select `red` as color:

```
2019-08-13 08:27:44.491 [ome.event.ItemCommandEvent] - Item 'KChe1_Color' received command 0,100,100
2019-08-13 08:27:44.492 [nt.ItemStatePredictedEvent] - KChe1_Color predicted to become 0,100,100
2019-08-13 08:27:44.493 [vent.ItemStateChangedEvent] - KChe1_Color changed from 73,81,100 to 0,100,100
2019-08-13 08:27:53.537 [vent.ItemStateChangedEvent] - KChe1_ColorTemperature changed from 87 to NULL
```

outcome:
* temperature is off
* alexa shows `red` correctly

everything fine so far

---

* go back to "Warm white" temperature

```
2019-08-13 08:39:17.889 [ome.event.ItemCommandEvent] - Item 'KChe1_ColorTemperature' received command 86.66666666666667
2019-08-13 08:39:17.890 [nt.ItemStatePredictedEvent] - KChe1_ColorTemperature predicted to become 86.66666666666667
2019-08-13 08:39:17.891 [vent.ItemStateChangedEvent] - KChe1_ColorTemperature changed from NULL to 86.66666666666667
2019-08-13 08:39:25.094 [vent.ItemStateChangedEvent] - KChe1_Color changed from 0,100,100 to 73,81,100
2019-08-13 08:39:25.095 [vent.ItemStateChangedEvent] - KChe1_ColorTemperature changed from 86.66666666666667 to 87
```

outcome:
* The light bulb is fine. It's "warm white" as selected
* **Alexa App mixes up the color and color temperature and shows some (ugly) greenish custom color.**

---

I assume the only solution would be to *not* report the color when the temperature is set. But that feels like a ugly hue specific hack ... Not sure if it's worth reopening the issue.

**Comment by jsetton** [comment]
> I assume the only solution would be to not report the color when the temperature is set. But that feels like a ugly hue specific hack ... Not sure if it's worth reopening the issue.

So the problem is related to the color mode determination logic. The reporting part is actually in line with the current implementation.

Based on the research I did to resolve this one, I recall seeing other bindings (e.g. lifx) supporting colored lights with their temperature dimmer item-typed value not reset in color mode and would solely rely on the color saturation (`hue,saturation,brightness`) to determine that mode.

This is actually back firing in your case since the color saturation for hue light seems to be greater then zero in white mode, based on your example. I have to think about this one but I don't see how we can accommodate both scenarios unless we add an extra parameter.

https://github.com/openhab/openhab-alexa/blob/07dd252fe4504eef8a3aae8347e0654aeff0a2d8/lambda/smarthome/alexa/capabilities.js#L317-L330

**Comment by jsetton** [comment]
@nougad I just opened a PR that should fix the issue you mentioned. Any chance you can test, deploying your own skill, on your end to confirm it is working as expected?

**Comment by jsetton** [comment]
@nougad I went ahead and merged the PR so it would be available on the new skill rollout.
